### PR TITLE
Accept bytes for Unix socket paths.

### PR DESCRIPTION
**Pull request by Fuyukai**
*nix paths are an arbitrary sequence of bytes.

### CODE REVIEW

**Comment by agronholm** [comment]
While this is *almost* an annotations only patch, it does replace the use of `str()`. Perhaps there should be a test that creates a socket from a bytes-based path that doesn't cleanly translate to ASCII?

**Comment by Fuyukai** [comment]
What would that actually be testing? It's just passed straight through to the underlying socket.

**Comment by agronholm** [comment]
It would test that a bytes-based socket path would not result in `UnicodeDecodeError`, as it could without this patch.

**Comment by agronholm** [comment]
It's even clearer now that I tried:
```
>>> Path(b"/tmp/foo2")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.12/pathlib.py", line 1163, in __init__
    super().__init__(*args)
  File "/usr/lib64/python3.12/pathlib.py", line 373, in __init__
    raise TypeError(
TypeError: argument should be a str or an os.PathLike object where __fspath__ returns a str, not 'bytes'
```

**Comment by Fuyukai** [comment]
Also added bytes supports for the other methods and added tests for each distinct one.

File: src/anyio/_core/_sockets.py [hunk 1]
```diff
@@ -245,7 +247,9 @@ async def try_connect(remote_host: str, event: Event) -> None:
     return connected_stream
 
 
-async def connect_unix(path: str | PathLike[str]) -> UNIXSocketStream:
+async def connect_unix(
+    path: str | bytes | PathLike[str] | PathLike[bytes]
```
Comment (agronholm) [inline src/anyio/_core/_sockets.py:251]: Hm, wouldn't `PathLike[Any]` work here, given that the variable is bound to `str` and `bytes` anyway?

File: src/anyio/_core/_sockets.py [hunk 2]
```diff
@@ -340,7 +344,10 @@ async def create_tcp_listener(
 
 
 async def create_unix_listener(
-    path: str | PathLike[str], *, mode: int | None = None, backlog: int = 65536
+    path: str | bytes | PathLike[str] | PathLike[bytes],
```
Comment (agronholm) [inline src/anyio/_core/_sockets.py:347]: Same as above.

File: src/anyio/_core/_sockets.py [hunk 3]
```diff
@@ -466,7 +473,7 @@ async def create_connected_udp_socket(
 
 async def create_unix_datagram_socket(
     *,
-    local_path: None | str | PathLike[str] = None,
+    local_path: None | str | bytes | PathLike[str] = None,
```
Comment (agronholm) [inline src/anyio/_core/_sockets.py:476]: Shouldn't this also accept `PathLike[bytes]`?

File: tests/test_sockets.py [hunk 4]
```diff
@@ -1002,6 +1002,24 @@ async def test_cannot_connect(self, socket_path: Path) -> None:
...
             await connect_unix(socket_path)
 
+    async def test_connecting_using_bytes(
+        self, server_sock: socket.socket, socket_path: Path
+    ) -> None:
+        async with await connect_unix(str(socket_path).encode()):
+            pass
+
+    @pytest.mark.skipif(
+        sys.platform == "darwin", reason="OSX requires valid UTF-8 paths"
```
Comment (agronholm) [inline tests/test_sockets.py:1012]: Other AnyIO skip markers use `platform.system()`; this should remain consistent.
```suggestion
        platform.system() == "Darwin", reason="macOS requires valid UTF-8 paths"
```

File: tests/test_sockets.py [hunk 5]
```diff
@@ -1101,6 +1119,18 @@ async def test_bind_twice(self, socket_path: Path) -> None:
             async with await create_unix_listener(socket_path):
                 pass
 
+    async def test_listening_bytes_path(self, socket_path: Path) -> None:
+        async with await create_unix_listener(str(socket_path).encode()):
+            pass
+
+    @pytest.mark.skipif(
+        sys.platform == "darwin", reason="OSX requires valid UTF-8 paths"
```
Comment (agronholm) [inline tests/test_sockets.py:1127]: As above.
```suggestion
        platform.system() == "Darwin", reason="macOS requires valid UTF-8 paths"
```

File: tests/test_sockets.py [hunk 6]
```diff
@@ -1495,6 +1525,20 @@ async def test_send_after_close(self, socket_path: Path) -> None:
...
             await unix_dg.sendto(b"foo", path)
 
+    async def test_local_path_bytes(self, socket_path: Path) -> None:
+        async with await create_unix_datagram_socket(
+            local_path=str(socket_path).encode()
+        ):
+            pass
+
+    @pytest.mark.skipif(
+        sys.platform == "darwin", reason="OSX requires valid UTF-8 paths"
```
Comment (agronholm) [inline tests/test_sockets.py:1535]: Ditto.
```suggestion
        platform.system() == "Darwin", reason="macOS requires valid UTF-8 paths"
```

**Review by agronholm — CHANGES_REQUESTED** [review]

**Comment by agronholm** [comment]
Also please add a changelog entry in `docs/versionhistory.rst`.

File: src/anyio/_core/_sockets.py (hunk 1 above)
Comment (Fuyukai) [reply src/anyio/_core/_sockets.py:251]: Pylance gave me an error when I first tried, but Pyright doesn't. Not sure what's going on there.. I'll trust the CLI and change it.

Comment (agronholm) [reply src/anyio/_core/_sockets.py:251]: If the CI passes, then that's good enough for me.

File: tests/test_sockets.py (hunk 4 above)
Comment (Fuyukai) [reply tests/test_sockets.py]: For what it's worth, I copied it from the one immediately below which uses ``sys.platform`` :)

Comment (agronholm) [reply tests/test_sockets.py]: Oh, that's on me then :)
All the newer skip marks use `platform.system()`.

**Review by agronholm — APPROVED** [review]
LGTM now, thanks!

**Pull request merged**